kingpin/actors/packagecloud.py

```diff
@@ -0,0 +1,400 @@
+
+        Args:
+            name: name of the package to look for
+            all_packages: hash of all the packages, as returned by the API
+
+        Returns:
+            A hash of package versions sorted by creation date
+        """
+        versions = [{
+            'created_at': datetime.datetime.strptime(
+                package['created_at'], '%Y-%m-%dT%H:%M:%S.%fZ'),
+            'distro_version': package['distro_version'],
+            'filename': package['package_html_url'].split('/')[-1],
+            'name': package['name']
+        } for package in all_packages if package['name'] == name]
+
+        versions.sort(key=lambda x: x.get('created_at'), reverse=True)
+        raise gen.Return(versions)
+
+    @gen.coroutine
+    def _get_packages_list_to_delete(self, packages_to_delete, all_packages):
+        """Extracts a list of unique package names to delete
+
+        Args:
+            packages_to_delete: regex of package names to delete
+            all_packages: hash of all the packages, as returned by the API
+
+        Returns:
+            A list of unique package names that match the delete pattern.
+            """
+        pattern = re.compile(packages_to_delete)
+        packages_list_to_delete = [package['name'] for package in all_packages
+                                   if pattern.match(package['name'])]
+        packages_list_to_delete = set(packages_list_to_delete)
+        raise gen.Return(packages_list_to_delete)
+
+    @gen.coroutine
+    def _delete(self, packages_to_delete, repo, older_than=0,
+                number_to_keep=0):
+        """Generic packagecloud delete method, optionally supporting deleting
+        old packages by date and/or keeping a certain number of packages.
+
+        Args:
+            packages_to_delete: Regex of packages to delete, e.g. pkg1|pkg2
+            repo: name of the packagecloud repo to delete from
+            older_than: Delete packages created before this number of seconds
+            number_to_keep: Keep at least this number of each package
+
+        Returns:
+            A list of the packages that were deleted
+        """
+        all_packages = yield self._get_all_packages(repo=repo)
+        packages_list_to_delete = yield self._get_packages_list_to_delete(
+            packages_to_delete, all_packages)
+        deleted_packages = []
+
+        # Loop through each unique package to delete
+        for name in packages_list_to_delete:
+            package_versions = yield self._get_package_versions(
+                name, all_packages)
+
+            # Delete individual packages if they meet our criteria
+            number_in_repo = len(package_versions)
+            for package in package_versions:
+                if number_in_repo > number_to_keep:
+                    if older_than:
+                        if datetime.datetime.now() - package['created_at'] > \
+                           datetime.timedelta(seconds=older_than):
```

Comment:
I would create named values for these. What is `datetime.datetime.now() - package['created_at']` ? package_age would be great. Then the second part is allowed_age ? So your final if would say `if package_age > allowed_age` and that's very human readable